cpmpy/solvers/scip.py

```diff
@@ -0,0 +1,419 @@
+        return cpm_cons
+
+    def __add__(self, cpm_expr_orig):
+      """
+            Eagerly add a constraint to the underlying solver.
+
+            Any CPMpy expression given is immediately transformed (through `transform()`)
+            and then posted to the solver in this function.
+
+            This can raise 'NotImplementedError' for any constraint not supported after transformation
+
+            The variables used in expressions given to add are stored as 'user variables'. Those are the only ones
+            the user knows and cares about (and will be populated with a value after solve). All other variables
+            are auxiliary variables created by transformations.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: self
+      """
+      import pyscipopt as scip
+
+      # add new user vars to the set
+      get_variables(cpm_expr_orig, collect=self.user_vars)
+
+      # transform and post the constraints
+      for cpm_expr in self.transform(cpm_expr_orig):
+
+        # Comparisons: only numeric ones as 'only_bv_implies()' has removed the '==' reification for Boolean expressions
+        # numexpr `comp` bvar|const
+        if isinstance(cpm_expr, Comparison):
+            lhs, rhs = cpm_expr.args
+            sciprhs = self.solver_var(rhs)
+
+            # Thanks to `only_numexpr_equality()` only supported comparisons should remain
+            if cpm_expr.name == '<=':
+                if (isinstance(lhs, Operator) and lhs.name == "sum" and all(a.is_bool() and not isinstance(a, NegBoolView) for a in lhs.args)):
+                    if rhs == 1: # special SOS1 constraint?
+                        self.scip_model.addConsSOS1(self.solver_vars(lhs.args))
+                    else: # cardinality constraint
+                        self.scip_model.addConsCardinality(self.solver_vars(lhs.args), rhs)
+                else:
+                    sciplhs = self._make_numexpr(lhs)
+                    self.scip_model.addCons(sciplhs <= sciprhs)
+
+            elif cpm_expr.name == '>=':
+                sciplhs = self._make_numexpr(lhs)
+                self.scip_model.addCons(sciplhs >= sciprhs)
+            elif cpm_expr.name == '==':
+                if isinstance(lhs, _NumVarImpl) \
+                        or (isinstance(lhs, Operator) and (lhs.name == 'sum' or lhs.name == 'wsum' or lhs.name == "sub")):
+                    # a BoundedLinearExpression LHS, special case, like in objective
+                    sciplhs = self._make_numexpr(lhs)
+                    self.scip_model.addCons(sciplhs == sciprhs)
+
+                elif lhs.name == 'mul':
+                    assert len(lhs.args) == 2, "scip only supports multiplication with 2 variables"
+                    a, b = self.solver_vars(lhs.args)
+                    #self.scip_model.setParam("NonConvex", 2)  # gurobi leftover
```

Comment:
This can be removed for SCIP. SCIP allows many more non-linear type formulations than just quadratic.
If you want to give options to SCIP on how to handle quadratic constraints then you can change the existing parameters.